core/base/temporary_conversion.hpp

```diff
@@ -0,0 +1,227 @@
+            cast_obj->convert_to(converted.get());
+            static_assert(
+                std::is_base_of<ConvertibleTo<std::remove_cv_t<TargetType>>,
+                                FirstCandidate>::value,
+                "ConvertibleTo not implemented");
+            static_assert(std::is_base_of<ConvertibleTo<FirstCandidate>,
+                                          TargetType>::value,
+                          "ConvertibleTo not symmetric");
+            return {converted.release(),
+                    convert_back_deleter<TargetType, candidate_type>{cast_obj}};
+        } else {
+            return conversion_helper<TrailingCandidates...>::template convert<
+                TargetType>(obj);
+        }
+    }
+};
+
+template <>
+struct conversion_helper<> {
+    template <typename T, typename MaybeConstLinOp>
+    static std::unique_ptr<T, std::function<void(T *)>> convert(
+        MaybeConstLinOp *obj, bool initialize)
+    {
+        throw NotSupported(__FILE__, __LINE__,
+                           std::string{"gko::make_temporary_conversion<"} +
+                               name_demangling::get_type_name(typeid(T)) + ">",
+                           name_demangling::get_type_name(typeid(*obj)));
+        return {nullptr, null_deleter<T>{}};
+    }
+};
+
+
+/**
+ * A temporary_conversion is a special smart pointer-like object that is
+ * designed to hold an object temporarily converted to another format.
+ *
+ * After the temporary_conversion goes out of scope, the stored object will
+ * be converted back to its original format. This class is optimized to
+ * avoid copies if the object is already in the correct format, in which
+ * case it will just hold a reference to that object, without performing the
+ * conversion.
+ *
+ * @tparam T  the type of object held in the temporary_conversion
+ */
+template <typename T>
+class temporary_conversion {
+public:
+    using value_type = T;
+    using pointer = T *;
+    using lin_op_type = typename std::conditional<std::is_const<T>::value,
+                                                  const LinOp, LinOp>::type;
+
+    template <typename... Formats>
+    static temporary_conversion create(lin_op_type *ptr)
+    {
+        T *cast_ptr{};
+        if ((cast_ptr = dynamic_cast<T *>(ptr))) {
+            return handle_type{cast_ptr, null_deleter<T>{}};
+        } else {
+            return conversion_helper<Formats...>::template convert<T>(ptr);
+        }
+    }
+
+
+    /**
+     * Returns the object held by temporary_conversion.
+     *
+     * @return the object held by temporary_conversion
+     */
+    T *get() const { return handle_.get(); }
+
+    /**
+     * Calls a method on the underlying object.
+     *
+     * @return the underlying object
+     */
+    T *operator->() const { return handle_.get(); }
+
+    /**
+     * Returns if the conversion was successful.
+     */
+    explicit operator bool() { return bool(handle_); }
```

Comment:
nit: using C++ cast:
```suggestion
    explicit operator bool() { return static_cast<bool>(handle_); }
```